# Sprint 1 – Dataset Understanding

## Objective

The objective of this sprint is to understand the dataset before performing any preprocessing or machine learning.

A Machine Learning model is only as good as the data it learns from.

Therefore, before writing any model, we must inspect the dataset, understand every feature, identify potential problems, and verify data quality.

# Why Dataset Understanding?

Imagine a doctor prescribing medicine without checking the patient.

That would be dangerous.

Similarly, training a Machine Learning model without understanding the dataset often leads to poor models and incorrect conclusions.

Professional Data Scientists always inspect the dataset before building models.

In [1]:
# ==========================================
# Sprint 1 - Dataset Understanding
# ==========================================

import numpy as np 
import pandas as pd 

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [10]:
df = pd.read_csv("../data/ford.csv")
df.head()

,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,Fiesta,2017,12000,Automatic,15944,Petrol,150,57.7,1.0
1,Focus,2018,14000,Manual,9083,Petrol,150,57.7,1.0
2,Focus,2017,13000,Manual,12456,Petrol,150,57.7,1.0
3,Fiesta,2019,17500,Manual,10460,Petrol,145,40.3,1.5
4,Fiesta,2019,16500,Automatic,1482,Petrol,145,48.7,1.0


In [11]:
df.shape

(17966, 9)

In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 17966 entries, 0 to 17965
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         17966 non-null  str    
 1   year          17966 non-null  int64  
 2   price         17966 non-null  int64  
 3   transmission  17966 non-null  str    
 4   mileage       17966 non-null  int64  
 5   fuelType      17966 non-null  str    
 6   tax           17966 non-null  int64  
 7   mpg           17966 non-null  float64
 8   engineSize    17966 non-null  float64
dtypes: float64(2), int64(4), str(3)
memory usage: 1.2 MB


# Sprint 1 - Part 2

## Objective

The objective of this section is to inspect the quality of the dataset before applying any preprocessing or machine learning algorithms.

Data quality directly affects model performance. Therefore, we must verify that the dataset is complete, consistent, and suitable for analysis.

In [13]:
df.describe()

,year,price,mileage,tax,mpg,engineSize
count,17966.000000,17966.000000,17966.000000,17966.000000,17966.000000,17966.000000
mean,2016.866470,12279.534844,23362.608761,113.329456,57.906980,1.350807
std,2.050336,4741.343657,19472.054349,62.012456,10.125696,0.432367
min,1996.000000,495.000000,1.000000,0.000000,20.800000,0.000000
25%,2016.000000,8999.000000,9987.000000,30.000000,52.300000,1.000000
50%,2017.000000,11291.000000,18242.500000,145.000000,58.900000,1.200000
75%,2018.000000,15299.000000,31060.000000,145.000000,65.700000,1.500000
max,2060.000000,54995.000000,177644.000000,580.000000,201.800000,5.000000


## Observation

Write 4–5 observations from describe().

Example

- Average price is 12279.53.
- Average mileage is 23362.
- Maximum engine size is 1.35.
- The dataset contains 17966 observations.

In [14]:
df.isnull().sum()

model           0
year            0
price           0
transmission    0
mileage         0
fuelType        0
tax             0
mpg             0
engineSize      0
dtype: int64

In [15]:
df.duplicated().sum()

np.int64(154)

In [17]:
df.nunique()

model              24
year               23
price            3511
transmission        3
mileage         13528
fuelType            5
tax                36
mpg                90
engineSize         16
dtype: int64

In [18]:
df.memory_usage(deep=True)

Index              132
model           998222
year            143728
price           143728
transmission    995474
mileage         143728
fuelType        988133
tax             143728
mpg             143728
engineSize      143728
dtype: int64

| Feature | Data Type | Description |
|----------|-----------|-------------|
| model | Categorical | Name of Ford model |
| year | Numerical | Manufacturing year |
| price | Target | Selling price |
| mileage | Numerical | Distance travelled |
| fuelType | Categorical | Fuel type |
| transmission | Categorical | Transmission type |
| tax | Numerical | Road tax |
| mpg | Numerical | Miles per gallon |
| engineSize | Numerical | Engine size |

# Sprint Summary

Today I learned:

- How to inspect dataset quality.
- How to identify missing values.
- How to detect duplicate rows.
- How to understand descriptive statistics.
- How to create a Data Dictionary.